Helper script for performing controller forced reconfiguration

joe-redpanda opened the pull request:
Adds a basic orchestrator for controller forced reconfiguration.

Usages:
1. python3 tools/controller_forced_reconfiguration.py baremetal --nodes IP1 IP2 IP3 --dead-nodes 1 2 3 --surviving-count 2
2. python3 tools/controller_forced_reconfiguration.py baremetal --nodes IP1 IP2 IP3 --check-liveness
3. python3 tools/controller_forced_reconfiguration.py kubernetes

1: baremetal manual mode
simply muxes the dead nodes and surviving node count into controller forced reconfiguration commands to each provided ip. IPs should be the list of living nodes, parameters should match accordingly.

2: baremetal recommendation mode
The script will scrape every provided IP to try to generate a list of living nodes, dead nodes, and then transmute those into controller forced reconfiguration parameters. The recommendation will only succeed if the cluster topology is trivial (all remaining living nodes can see each other, and the living node count matches the node status response from each node).

3: kubernetes
The script will leverage kubectl to orchestrate CFR. This requires the precondition that kubectl is logged into the cluster in question.

The script will gather the list of redpanda pods from kubectl, setup tunnels to contact their admin apis, and then proceed with a liveness check orchestration as listed above.

The same rules apply, the script will only attempt CFR for a trivial cluster topology.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

joe-redpanda committed:
controller forced reconfiguration helper script:

Adds a helper script to ease the process of invoking CFR against a
number of redpanda node ips.

Can do three things
1. mux a set of controller forced reconfiguration parameters to all
   living nodes
2. given a set of ips, recommend and execute those controller forced
   reconfiguration parameters
3. given a kubernetes cluster, gather all redpanda endpoints, recommend
   controller forced reconfiguration parameters, and execute with those
   parameters

joe-redpanda commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,340 @@
+                "success": response.status_code in [200, 201, 204]
+            }
+
+            if self.result["success"]:
```
This needs an additional check, CFR can succeed with a 4xx if one of the nodes wins the election before all CFR requests have landed. This is normal / expected and should probably just be logged out

joe-redpanda committed:
type-check-strictness.json: ignore CFR script

Skips all type checking on the CFR orchestrator script.

joe-redpanda commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+
+    detail:
+        For each provided node, gather that node's perspective on every other node's liveness to check whether we can naiively apply CFR.
+        We can naiively perform CFR if we, the operator, can reach all living nodes, and all living nodes can reach eachother.
```
one day i'll learn how to spell

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+
+    2. try to generate a command
+    python invoke_controller_reconfiguration.py --nodes IP1 IP2 IP3 --check-liveness
+     - the script will call request node status information from all provided ips (usually the ips of all nodes alive or dead)
```
nit: IPs?

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+    first_alive_set = alive_node_sets[0]
+    all_agree = all(alive_set == first_alive_set for alive_set in alive_node_sets)
+
+    # Get all node IDs from any successful response
```
nit: from **all** successful response**s**?

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+    print(f"{'=' * 70}")
+
+    successful_checkers = [c for c in checkers if c.alive_nodes is not None]
+    failed_checkers = [c for c in checkers if c.error is not None]
```
isn't a checker always either failed or successful but not both?

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+    # Get all node IDs from any successful response
+    all_node_ids = set()
+    for checker in successful_checkers:
+        if checker.brokers_data:
```
It only can evaluate to false if it's empty (and hence they all are). I assume it's not possible, as each node returns at least itself?

joe-redpanda replied:
successful implies that brokers_data is truthy, Type annotations is off for this file, but it would complain if enabled. I could take or leave the if check, your call

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+    print(f"{'=' * 70}")
+
+    # Does the size of the living node set match the number of responses
+    living_nodes_are_reachable = len(successful_checkers) == len(first_alive_set)
```
We don't attempt to match IPs or cluster IDs, just check counts. In theory we might have nodes 1 and 2 from one cluster that has nodes 1-5 alive, and nodes 3-5 from another unrelated cluster that also has nodes 1-5 alive. Should we guard from such a situation?

joe-redpanda replied:
I'm not terribly worried about that one because this is atm a support only script.

What would happen in that case is the CFR would stall on the nodes from the target cluster (cfr is volatile until success so just restart them) and be rejected on the oops cluster (assuming that cluster is healthy)

This does raise the concern of what happens when a cluster is experiencing a network partition:
lets say 5 node cluster split into 2, 2, and 1 s.t. no chunk has quorum.

CFR would stall on all nodes, and nothing would happen.

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+    if args.check_liveness:
+        liveness_result = check_cluster_liveness(args.nodes, args.port)
+
+        if liveness_result is None or not liveness_result.get("consensus"):
```
why `.get`, can it really be non-null and not a dict with`consensus` key?

joe-redpanda replied:
Yeah I'll do a cleanup pass before I rerequest review

I claude generated the shape of the script and then scrubbed for correctness and documentation, but I didn't pick through for these kind of oddities.

bashtanov reviewed (commented):
LGTM, a few questions

daisukebe commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+        self.alive_nodes = None
+
+    def check(self):
+        url = f"http://{self.ip}:{self.port}/v1/brokers"
```
In a case the admin endpoint is secured (TLS enabled), `https` should be used. It is a per-broker configuration, but I guess it's better to be controlled by a flag in this command. Wdyt?

joe-redpanda replied:
Thats a good point, let me ask around about tls and gather some info

daisukebe commented:
This script appears to be designed for bare-metal or self-hosted k8s deployments only. There's a blocker in our cloud deployment to manage to run it. Let's discuss about our goal.

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python3
+"""
+Helper script to orchestrate invoking Controller Forced Reconfiguration.
```
a high level question. Users are expected to invoke this via `rpk`, would it be easier to implement there for simpler integration?

joe-redpanda replied:
This is not meant to be invoked via RPK

joe-redpanda replied:
distilling the discussion:

internal only apis don't get generated into go bindings for rpk && this endpoint isn't really meant to be invoked by customers.

So we were left with an odd api that required orchestration to work reasonably but no ability to put that in go.

This is (afaik) the only api like this so we made the judgement call
unique api -> one off solution

bharathv replied:
> This is not meant to be invoked via RPK

I’m not sure I agree on this. Keeping things consistent in rpk feels important, and we already have similar commands like unsafe-recover there, so there’s some precedent. Even if it’s a risky operation, I think it should still live in the same CLI across all modes (k8s, SH etc) the rest is just implementation details (making sure the right bindings are in place, etc.). Maybe we should check in with Matt again before going down this route.

Having this as a separate python file also brings up extra packaging questions, like how we include it in the release artifacts and all that.

daisukebe replied:
My 2 cents:

I think we need a clear policy for developing this sort of feature. The unsafe-recover command invokes a non-debug endpoint (/v1/partitions/force_recover_from_nodes), so RPK integration was part of the [scope](https://docs.google.com/document/d/1n83PvnxYR8oJdTrlHkVozoTBsVelSNA4diQ3-3AMqwc/edit?tab=t.0#heading=h.5zvj0etnl8iy) from the beginning, or at least that's my understanding. On the other hand, we've generally avoided exposing /v1/debug/... endpoints to end users through RPK, which is why no such RPK command exists.

The question we need to answer with Matt is: does "internal" or "debug" mean we shouldn't develop a corresponding RPK command? If it depends on the situation, where do we draw the line?

joe-redpanda replied:
> does "internal" mean we shouldn't develop a corresponding RPK command

Currently, yes. Internal v2 admin apis do not get built into Go bindings and in lieu of a change to that cannot be added to RPK. If it's internal it's not in RPK.

This is also the case for shadow-linking escape hatches.

Will discuss the existing UX and other concerns with Matt. I'll share any novel findings

joe-redpanda commented:
> This script appears to be designed for bare-metal or self-hosted k8s deployments only. There's a blocker in our cloud deployment to manage to run it. Let's discuss about our goal.

Stellar, thank you for lending your eyes. I'll set up a discussion

daisukebe commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,772 @@
+                    # First column is the pod name
+                    pod_name = line.split()[0]
+                    # Filter for pods starting with 'rp-'
+                    if pod_name.startswith("rp-"):
```
This means the `kubernetes` commandt is supposed to be executed on our cloud deployments only? Users' self-hosted k8s deployment typically has different prefix.

joe-redpanda replied:
Any consistent naming or identification scheme, or should this be just ingesting a list of pod names?

daisukebe replied:
Unfortunately it varies in the wild in users' deployments 😓

daisukebe commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,772 @@
+        self.alive_nodes = None
+
+    def check(self):
+        url = f"http://{self.endpoint.ip}:{self.endpoint.port}/v1/brokers"
```
Consider supporting both http and https schemes based on user input, since some clusters may have TLS enabled on the Admin API.

joe-redpanda replied:
cluster wide or can tls be per-node

daisukebe replied:
It is per-node configuration, but it is consistent across nodes as long as the cluster is running fine. So we can assume TLS enabled on all nodes OR TLS disabled entirely.

bharathv replied:
also auth, I believe k8s clusters have some user/pass auth, don't they?

joe-redpanda replied:
yeah will track in follow on

daisukebe commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,772 @@
+            print(f"[{self.endpoint.ip}] ✗ Timeout error (node may be down)")
+        except requests.exceptions.ConnectionError as e:
+            self.error = f"Connection error: {str(e)}"
+            print(f"[{self.endpoint.ip}] ✗ Connection error (node may be down)")
```
`node may be down` --> `node may be down or https is required?`

When hitting `http` whereas TLS is configured, the script just reports `Connection error (node may be down)`, which is not accurate.

joe-redpanda replied:
Yeah, I'll update. I'm going to cut the script mostly as is for the moment and add some notes on the TLS and kubernetes limitation to the documentation alongside some best effort mitigation.

If TLS is heterogenous per node, it likely cannot be helped. At the point at which we have per node special things, we should just hit the node with the api manually

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+#!/usr/bin/env python3
```
(enterprise) license

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+    node_addresses: List[RedpandaAdminEndpoint],
+) -> Optional[Dict]:
+    """
+    Goal: can we naiively compile a CFR command to forcibly recover a cluster
```
nit: typo naively

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+        - responding_endpoints: list of RedpandaAdminEndpoint that responded successfully
+        - consensus: True if the we have full agreement per the detail note
+    """
+    print(f"\n{'=' * 70}")
```
nit: use a logger?

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+        )
+
+    def invoke(self):
+        url = f"http://{self.ip}:{self.port}/redpanda.core.admin.internal.v1.BreakglassService/ControllerForcedReconfiguration"
```
nit: use a constant for the suffix.

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+    port_forwards = []
+    local_port = starting_port
+
+    for pod in redpanda_pods:
+        cmd = [
+            "kubectl",
+            "port-forward",
+            f"pod/{pod}",
+            f"{local_port}:{admin_port}",
+            "--namespace",
+            namespace,
+        ]
```
Is this tested? I wonder if there is a library we could use

joe-redpanda replied:
maybe duplicate comment:
There is a kubernetes python lib, I intentionally didn't use that s.t. we could avoid dependency management.

This script is essentially automated documentation. It executes the commands that I otherwise would have asked support to execute in the appropriate order using the tools that you otherwise would already have on your machine in a support scenario.

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+            sys.exit(1)
+
+    print(f"\nSuccessfully started {len(port_forwards)} port forwards")
+    print(f"{'=' * 70}\n")
```
I'm curious what a sample output looks like :)

joe-redpanda replied:
documented [here](https://redpandadata.atlassian.net/wiki/spaces/CORE/pages/1475575854/Controller+Forced+Reconfiguration) with output

heres the relevant snippet though
```
joemiller@joe-box:~/git/dr/redpanda$ python3 tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py  kubernetes
Running: kubectl get pods --namespace redpanda
NAME                                        READY   STATUS      RESTARTS        AGE
redpanda-console-6c8d68cc58-bj8nq           1/1     Running     0               13m
rp-d5g35c4bleqoc5e7cjjg-blue-a-0            0/1     Running     1 (4m28s ago)   24m
rp-d5g35c4bleqoc5e7cjjg-blue-a-1            0/1     Running     1 (4m23s ago)   23m
rp-d5g35c4bleqoc5e7cjjg-blue-a-2            0/1     Running     0               112s
sasl-user-creation-first-superuser-9l65j    0/1     Error       0               22m
sasl-user-creation-first-superuser-fjg2t    0/1     Error       0               23m
sasl-user-creation-first-superuser-nqmhk    0/1     Completed   0               17m
sasl-user-creation-first-superuser-sntm7    0/1     Error       0               24m
sasl-user-creation-first-superuser-vb6xf    0/1     Error       0               20m
sasl-user-creation-first-superuser-xhtbr    0/1     Error       0               23m
sasl-user-creation-redpanda-console-9bn9p   0/1     Error       0               24m
sasl-user-creation-redpanda-console-dbjn4   0/1     Completed   0               18m
sasl-user-creation-redpanda-console-dg6qc   0/1     Error       0               21m
sasl-user-creation-redpanda-console-h4pqh   0/1     Error       0               24m
sasl-user-creation-redpanda-console-rlt5x   0/1     Error       0               23m
sasl-user-creation-redpanda-console-xz26t   0/1     Error       0               24m
Found 3 Redpanda cluster node pods:
  - rp-d5g35c4bleqoc5e7cjjg-blue-a-0
  - rp-d5g35c4bleqoc5e7cjjg-blue-a-1
  - rp-d5g35c4bleqoc5e7cjjg-blue-a-2
Are these the correct Redpanda cluster node pods? (yes/no): yes
======================================================================
Setting up port forwarding
======================================================================
Starting port forward: localhost:27900 -> rp-d5g35c4bleqoc5e7cjjg-blue-a-0:9644
  Ready: Forwarding from 127.0.0.1:27900 -> 9644
Successfully established port forward for rp-d5g35c4bleqoc5e7cjjg-blue-a-0
Starting port forward: localhost:27901 -> rp-d5g35c4bleqoc5e7cjjg-blue-a-1:9644
  Ready: Forwarding from 127.0.0.1:27901 -> 9644
Successfully established port forward for rp-d5g35c4bleqoc5e7cjjg-blue-a-1
Starting port forward: localhost:27902 -> rp-d5g35c4bleqoc5e7cjjg-blue-a-2:9644
  Ready: Forwarding from 127.0.0.1:27902 -> 9644
Successfully established port forward for rp-d5g35c4bleqoc5e7cjjg-blue-a-2
```


it just binds starting at 27900 to the admin port on the remote host
annoyingly, it runs this as a blocking operation hence the threads

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+
+    # Exit with error code if any requests failed
+    if failure_count > 0:
+        sys.exit(1)
```
probably deserves a log explaining why it exits.

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,781 @@
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Invoke ControllerForcedReconfiguration API on multiple Redpanda nodes concurrently"
```
I think some of the description on the top can be added in help text (when to run this command)

joe-redpanda commented:
basically a clean rewrite.

1. cleanup
    - extract functionality into common sense functions
    - consistency with exit criteria (usually an exception but a few cases where a result made more sense)
    - extract constants into actual constants
2. replaced prints with logger
3. substantially cut down the amount of logging and output
4. added basic auth
5. added tls
6. kubernetes will now respect the port option provided given that the default admin api cannot have tls enabled upon it
7. kubernetes can now configure the local port start (incase theres a conflict)


The last thing that is likely needed is an orchestration mode where the user can provide their own list of pods onto which to run CFR.

joe-redpanda commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,1128 @@
+# by the Apache License, Version 2.0
+
+newline = "\n"
+tab = "\t"
```
depending on the python version there are problems with having string literal escape sequences inside of a string thats being formatted

example:

```python
f"something something{"\n".join(my_list_to_be_joined_with_newlines)} something else"
```

joe-redpanda commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,1128 @@
+                pf.terminate(timeout=2.0)
+            except Exception as e:
+                # Log errors during cleanup for debugging
+                logger.error(
```
leaked resources if they don't close. Its irritating to have an invocation of CFR fail for whatever reason and then have to ss all of the ports in flight and kill the zombie processes

joe-redpanda commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,1128 @@
+    # print out CFR params and criteria such that the user can make an informed decision, then ask for confirmation
+    _invoke_cfr_confirmation(cfr_params)
+
+    # Convert sets to lists for compatibility with CFRInvoker
```
I'll clean this up if it gets another revision

joe-redpanda commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,1128 @@
+    return redpanda_pods
+
+
+# asks a user whether they want to continue or exit
```
will kick this into a
```
'''comment'''
```

style documentation if this gets another revision

bashtanov commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,1125 @@
+        if self.ca_cert:
+            args["verify"] = self.ca_cert
+        else:
+            args["verify"] = True  # Use system CA bundle
```
nit: can be simplified as `args["verify"] = self.ca_cert or True`

bharathv commented on tools/controller_forced_reconfiguration/invoke_controller_reconfiguration.py:
```diff
@@ -0,0 +1,1125 @@
+_active_port_forwards: list[PortForward] = []
+
+
+def cleanup_port_forwards() -> None:
+    """Clean up all active port forwards on exit."""
+    global _active_port_forwards
+    if _active_port_forwards:
+        logger.debug("cleaning up port forwards")
+        for pf in _active_port_forwards:
+            try:
+                pf.terminate(timeout=2.0)
+            except Exception as e:
+                # Log errors during cleanup for debugging
+                logger.error(
+                    f"failed to cleanup port forward - port: {pf.local_port}, "
+                    f"PID: {pf.process.pid}, "
+                    f"error: {str(e)}"
+                )
+        _active_port_forwards = []
+        logger.debug("cleaned up port forwards")
+
+
+def signal_handler(signum: int, _) -> None:
+    """Handle interrupt signals by cleaning up port forwards and exiting."""
+    logger.debug(f"signal {signum} caught")
+    cleanup_port_forwards()
+    sys.exit(1)
+
+
+# Register cleanup handlers
+atexit.register(cleanup_port_forwards)
+signal.signal(signal.SIGINT, signal_handler)
+signal.signal(signal.SIGTERM, signal_handler)
```
I think a cleaner way to do this by avoiding signal handlers is to use the context_manager pattern, yield a list of port_forwards and close them in finally, no mucking with signal handlers and no running signal unsafe stuff at exit etc.

joe-redpanda replied:
I'll do some destructor stuff in python to have it cleanup on close

bharathv reviewed (commented):
generally lgtm modulo one suggestion

hard to guarantee it works without some tests, perhaps something as simple as a running it against a ducktape cluster with various forms of auth-n helps acts as a regression tests.. k8s can be YOLO-ed tho.

joe-redpanda commented:
> generally lgtm modulo one suggestion
> 
> hard to guarantee it works without some tests, perhaps something as simple as a running it against a ducktape cluster with various forms of auth-n helps acts as a regression tests.. k8s can be YOLO-ed tho.

I don't disagree, but I think this arrives at 'evil rpk' as tyler had forewarned: a python rpk clone for core / internal only orchestration

I'll look into how to shove this script into a ducktape test. I'll probably have to do a little packaging magic to drop this into the ducktape image, but yeah it shouldn't be too too much different from the bog standard cfr tests

joe-redpanda commented:
Approved with the condition of adding a ducktape integration for this later. I'll take the point of adding some simple e2e CFR tests with basic auth, certs, and a few others. The trick is going to be shoving the cfr script into the ducktape image.

This has been manually tested for the time being
